### PR TITLE
Fix #1 - Added fireplace support and much more

### LINKED ISSUE #1: [Feature Request] - Add support for fireplaces
Add support for fireplaces http://docs-local.appbond.com/#tag/Devices

### PR DESCRIPTION
**Pull request by niemyjski**
_Please note: This is my first time really writing python so any feedback or pointers are very very welcome! I've been a professional programmer for a long time and felt it was pretty simple to pick up but I'm still learning._

### Remaining Items
- [x] I need the request library pr to be pushed before I can test: https://github.com/nguyer/bond-home/pull/2
- [x] Test these changes

### Completed Items
- [x] Added Fireplace Support as a brightness light controlling the flame. I had considered a climate device but found the increase flame part didn't seem to fit in.
- [x] pep8 compliance
- [x] Synced constants and api calls from the updated bond pr, and removed any constant imports we were not using.
- [x] Added `location` to `name` prefix.
- [x] Start reporting device properties, device id and unique id
- [x] I was looking into what was required for async support down the road, one of the items is not doing async inside of `__init__`. With that said and I don't think it's generally a good practice to do long running calls inside of a constructor, I moved these out into setup which will work really nicely once we do async support.

### CODE REVIEW

File: custom_components/bond/light.py [hunk 1]
```diff
@@ -1,53 +1,74 @@
...
 class BondLight(Light):
     """Representation of an Bond Light."""
 
-    def __init__(self, bond, deviceId):
+    def __init__(self, bond, deviceId, properties):
         """Initialize a Bond Light."""
         self._bond = bond
         self._deviceId = deviceId
-        self._properties = self._bond.getDevice(self._deviceId)
-        self._name = self._properties['name'] + ' Light'
```
Comment (niemyjski) [inline custom_components/bond/light.py]: we are already in the light domain. It makes no sense to prefix Light.

**Comment by niemyjski** [comment]
@bcb2000, @marciogranzotto, @nguyer, @a200462790, @Timp103 Can you please review this pr!

File: custom_components/bond/light.py [hunk 2]
```diff
@@ -22,71 +33,151 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
...
+            # Ceiling Fan but has no action for light control
+            # then don't create a light instance.
+            supportsLightActions = \
+                BOND_DEVICE_ACTION_TURN_LIGHT_ON in actions or \
+                BOND_DEVICE_ACTION_TURN_LIGHT_OFF in actions or \
+                BOND_DEVICE_ACTION_TOGGLE_LIGHT in actions
+            if supportsLightActions:
+                deviceProperties = self._bond.getProperties(deviceId)
+                light = BondLight(bond, deviceId, device, deviceProperties)
+                add_entities([light])
```
Comment (niemyjski) [inline custom_components/bond/light.py:51]: I wonder if it's better to build up a list and call this once

**Comment by nguyer** [comment]
Thanks for opening this! I’ll try to review it this weekend!

**Comment by niemyjski** [comment]
@nguyer if you could review and get the bond-home client released (sooner the better). I have time this morning to test it in HA.

**Comment by niemyjski** [comment]
@nguyer Thanks! If you could please review and get the bond-home client released (sooner the better). I have time this morning to test this pr in HA.

File: custom_components/bond/fan.py [hunk 3]
```diff
@@ -110,3 +108,13 @@ def update(self):
         bondState = self._bond.getDeviceState(self._deviceId)
         if 'power' in bondState:
             self._state = True if bondState['power'] == 1 else False
+
+    @property
+    def unique_id(self):
+        """Get the unique identifier of the device."""
+        return self._deviceId
```
Comment (nguyer) [inline custom_components/bond/fan.py:115]: This is a good addition! Thanks!

File: custom_components/bond/light.py (hunk 2 above)
Comment (nguyer) [reply custom_components/bond/light.py:51]: Probably. I got that impression, reading the docs when I first wrote this, but this implementation was quicker at the time, and I had no idea what I was doing in Python...

Feel free to implement it as a single call, or we can get it later in another update.

File: custom_components/bond/light.py [hunk 4]
```diff
@@ -22,71 +31,191 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
...
+        return self._deviceId
+
+    @property
+    def device_id(self):
+        """Return the ID of this light."""
+        return self.unique_id
+
+
+class BondFireplace(Light):
+    """Representation of an Bond Fireplace."""
```
Comment (nguyer) [inline custom_components/bond/light.py:128]: I guess this makes sense to treat a fireplace like a light. I get it, with the dimmer and all... it's just kinda funny to think about. I guess it does emit light...

**Review by nguyer — APPROVED** [review]
This is great! I don't have a fireplace to test this out, but I'll trust your testing :) Have you had a chance to run this branch on a real HA install?

I'll get https://github.com/nguyer/bond-home/pull/2 merged and published before merging this one.

File: custom_components/bond/light.py (hunk 2 above)
Comment (niemyjski) [reply custom_components/bond/light.py:51]: I'll pass on this for now, but we should get it in a later update.

File: custom_components/bond/light.py (hunk 4 above)
Comment (niemyjski) [reply custom_components/bond/light.py:128]: Yeah, I thought the same. There really is no other entity type that fits in here though :\

**Comment by niemyjski** [comment]
@nguyer I've been running this in my HA Prod instance for 3 days now. The only bug I've really ran into is with the fireplace (I haven't been able to test the other types but they do not through errors during setup). Here's the issue and it's going to have to be solved at some point but I'm not sure how...

The HA has a brightness value of 0-255 and the bond controller has a brightness of 0-100. Problem is all devices seem to operate differently like I saw the comment for the hard coded fan having 3 levels. I can definitely see more than that but at the same time, every fan I've ever owned has had 3. For example, my fireplace has a flame up and flame down and calling SetFlame doesn't do anything useful it just triggers the RF call with the converted brightness level. BUT how my fireplace actually works is it moves the flame valve up and down while HOLDING the RF signal... So I'm really really not sure how to get brightness to work reliably as well as get the max flame level as bond doesn't know that and I don't know how they would.

TLDR; Brightness is there on the fireplace but probably isn't going to work the greatest...

**Comment by nguyer** [comment]
> The HA has a brightness value of 0-255 and the bond controller has a brightness of 0-100.

Okay great. Makes sense. We probably just need to divide the HA brightness by 2.55 and round, to map it to the bond controller brightness. I would assume that brightness is always in the range 0 - 100 in the Bond API. This is true of my dimmable fan lights. Though maybe that's not a safe assumption. I'm trying to look it up in the docs, but it appears the doc site is down right now.

> Problem is all devices seem to operate differently like I saw the comment for the hard coded fan having 3 levels. I can definitely see more than that but at the same time, every fan I've ever owned has had 3.

I think this is actually a separate thing, and I think it's already addressed by the fact that fan devices report their maximum speed as an attribute.

**Comment by socbrian** [comment]
@nguyer are you going to merge this branch into the main one or keep this separate?

**Comment by niemyjski** [comment]
I actually did the math to convert it, several other integrations do it already as part of the core repo

**Comment by nguyer** [comment]
> @nguyer are you going to merge this branch into the main one or keep this separate?

I have a couple more things I want to test and possibly tweak in the `feature/fan-speed` branch, and then I'll merge it to master. Still can't change the speed of my fans from the dropdown in lovelace. It's not an issue with this PR though.

**Comment by niemyjski** [comment]
Also if someone wants to cache the get devices and get device info on init and store it on the domain object, we could save a boatload of time on startup 4-10 seconds

**Review by ViolentPoo — APPROVED** [review]

**Comment by niemyjski** [comment]
I've been running these changes for a few weeks without any issues. We should get this merged
